src/contexts/SwapProvider/swap-utils.ts

```diff
@@ -186,17 +222,90 @@ export async function ensureAllowance({
     }),
   );
 
-  if (signError) {
-    throwError(signError);
+  if (isUserRejectionError(signError)) {
+    throw signError;
+  } else if (signError) {
+    throw swapError(CommonError.UnableToSign, signError);
   }
 }
 
-export const throwError = (err: string | unknown): never => {
+const normalizeError = (err: unknown) => {
+  if (isWrappedError(err)) {
+    return err;
+  }
+
+  if (err instanceof Error) {
+    return err;
+  }
+
   if (typeof err === 'string') {
-    throw new Error(err);
+    return new Error(err);
   }
 
-  throw err;
+  return new Error((err as any)?.message ?? 'Unknown error');
+};
+
+export const swapError = (
+  errorCode: CommonError | SwapErrorCode,
+  originalError?: unknown,
+) => {
+  if (isWrappedError(originalError)) {
+    return originalError;
+  }
+
+  return ethErrors.rpc.internal({
+    data: {
+      reason: errorCode,
+      originalError: originalError
+        ? normalizeError(originalError)
+        : new Error('Unknown swap error'),
+    },
+  });
+};
+
+export const paraswapErrorToSwapError = (error: WrappedError): SwapError => {
+  if (!error.data.originalError) {
+    return {
+      message: t('Unknown error occurred, '),
+      hasTryAgain: true,
+    };
+  }
+
+  const originalError = error.data.originalError as Error;
+
+  switch (originalError.message) {
+    case 'ESTIMATED_LOSS_GREATER_THAN_MAX_IMPACT':
+      return {
+        message: t(
+          'Amount too low or too big to cover. Please adjust swap values.',
+        ),
+        hasTryAgain: false,
+      };
+
+    case 'No routes found with enough liquidity':
+      return {
+        message: t('No routes found with enough liquidity.'),
+        hasTryAgain: false,
+      };
+
+    case 'Internal Error while computing the price':
+      return {
+        message: t('An error occurred while computing the price.'),
+        hasTryAgain: false,
+      };
+  }
+
+  if (/is too small to proceed/.test(originalError.message)) {
+    return {
+      message: t('Amount is too small to proceed.'),
+      hasTryAgain: false,
+    };
+  }
+
+  return {
+    message: t('Unknown error occurred, '),
```

Comment:
Here as well? 😅

Reply:
Nah, it's the only case where `hasTryAgain` is true -- and when it is, a `try again` will be appended and result will be something like:

```Unknown error occurred, <button>try again</button>.```

Not exactly pretty, but I didn't want to do unnecessary refactoring in this PR, just improved the messages + started logging errors, and that's it :)